### PR TITLE
 Continuous highlighting of syntax errors #134 

### LINKED ISSUE #134: Continuous highlighting of syntax errors
Today, students have to click Run to realize they have made an obvious syntax error.  Would be great to find them and place obvious annotations on the editor, so that they can be fixed immediately without a context switch and having to go back and re-understand the syntax.

### PR DESCRIPTION
**Pull request by nixorn**
Closes #134

### CODE REVIEW

**Comment by nixorn** [comment]
In issue #134 you mentioned about priority system, which should control flood of requests to `compile` and `errorCheck` routes, because `errorCheck` makes same compilation, but triggered by `changes` CodeMirror event. And there could be excess server load. It is not implemented in this PR. Maybe we need to make separate issue for this.

**Comment by cdsmith** [comment]
Thanks!  I'll review this over the weekend.  I'm attending a Haskell hackathon tomorrow afternoon, so I'll probably make it my goal to get this merged and do all the UI tweaks during the hackathon, if you don't mind my jumping in.

File: web/js/codeworld.js [hunk 1]
```diff
@@ -160,7 +160,22 @@ function initCodeworld() {
             "Ctrl-Up": changeFontSize(1),
             "Ctrl-Down": changeFontSize(-1)
         },
-        textHover: onHover
+        textHover: onHover,
+        gutters: ["CodeMirror-lint-markers"],
+        lint: {
+            getAnnotations: function(text, callback, cm){
```
Comment (cdsmith) [inline web/js/codeworld.js]: cm seems to be unused here.  In JavaScript you can drop unused parameters from the end of a list.

**Comment by cdsmith** [comment]
Have you looked into interactions between this and doc-on-hover?  That is, I wonder what happens if there's an error on a symbol, but it's also a known variable.  Hopefully you wouldn't get two popups at once!  I suspect the right answer is that you should see the error popup, and not the documentation one.

**Comment by cdsmith** [comment]
I mentioned before that I'd like to get this merged during a Haskell hackathon today; but it looks like you may have local changes that you haven't pushed?

**Comment by nixorn** [comment]
Yes, I have some changes, but not finished yet. I plan to push them in hour or two.

**Comment by cdsmith** [comment]
Okay, great.  Thanks!

File: web/js/codeworld.js [hunk 2]
```diff
@@ -1016,3 +1031,41 @@ function downloadProject() {
...
+function parseCompileErrors(rawErrors){
+    var errors = [];
+    rawErrors = rawErrors.split("\n\n");
+    rawErrors.forEach(function(rawError){
+        rawError = rawError.split('\n');
+        var error = {},
+            annotation = rawError[0].trim(),
+            message = rawError.slice(1).map((err) => {return err.trim()}).join('\n');
+
+        var match = /^program.hs:(\d+):(\d+): (\w+):(.*)/.exec(annotation);
```
[the remark this answers is not in the document]
Comment (nixorn) [reply web/js/codeworld.js]: Added +1 pattern from https://github.com/google/codeworld/blob/master/web/js/codeworld_shared.js#L518 .

File: web/js/codeworld.js [hunk 3]
```diff
@@ -160,7 +160,22 @@ function initCodeworld() {
...
+        lint: {
+            getAnnotations: function(text, callback, cm){
+                var data = new FormData();
+                data.append("source", text)
+                var request = sendHttp(
+                    "POST", "errorCheck", data,
+                    function(request){
+                        if (request.status == 400){
+                            callback(parseCompileErrors(request.responseText))
+                        } else {callback([])};
```
[the remark this answers is not in the document]
Comment (nixorn) [reply web/js/codeworld.js]: Yes, this is error, thank you. Fixed.

File: web/js/codeworld.js [hunk 4]
```diff
@@ -1016,3 +1031,41 @@ function downloadProject() {
...
     }
 }
+
+function parseCompileErrors(rawErrors){
+    var errors = [];
+    rawErrors = rawErrors.split("\n\n");
+    rawErrors.forEach(function(rawError){
+        rawError = rawError.split('\n');
+        var error = {},
+            annotation = rawError[0].trim(),
```
[the remark this answers is not in the document]
Comment (nixorn) [reply web/js/codeworld.js]: Replaced.

File: web/js/codeworld.js [hunk 5]
```diff
@@ -1016,3 +1031,41 @@ function downloadProject() {
...
+                end = token.end;
+        } else {
+            var start = charNumber,
+                end = charNumber + 1;
+        }
+
+        error.from = CodeMirror.Pos(lineNumber, start);
+        error.to = CodeMirror.Pos(lineNumber, end);
+        error.severity = match[3];
+        if (match[4]){
```
[the remark this answers is not in the document]
Comment (nixorn) [reply web/js/codeworld.js]: Has been rewrote during adding new regexp pattern.

File: web/js/codeworld.js (hunk 1 above)
Comment (nixorn) [reply web/js/codeworld.js]: Thanks, removed.

**Comment by nixorn** [comment]
About interaction between this and doc-on-hover - doc tool-tip don't appear on underscored text. In most cases. Only exception is when underscored token is single token on line.
out

**Pull request merged**